lib/msal-browser/docs/acquire-token.md

```diff
@@ -89,10 +79,42 @@ var graphEndpoint = "https://graph.microsoft.com/v1.0/me";
+
+### Use a single PublicClientApplication instance
+
+Instantiate one `PublicClientApplication` per application and use the same instance throughout your app. This ensures that there is a single source of truth for what MSAL is performing at any given time (see: [MSAL events](events.md)) and eliminates the chance of distinct app objects making parallel interactive requests or potential cache conflicts, which might break apps, reduce performance or hinder user experience.
+
+### Always wait for promises to resolve
+
+All MSAL `acquireToken*` as well as `login*` APIs (which uses `acquireToken*` underneath) perform asynchronous operations and return promises. You should always wait for these promises to resolve, even if you do not need the promise payload.
+
+### Check for interaction status before interactive requests
+
+You should always check for [InteractionStatus](https://azuread.github.io/microsoft-authentication-library-for-js/ref/enums/_azure_msal_browser.interactionstatus.html) being equal to **None** before making an interactive request. `InteractionStatus` will be emitted through [MsalBroadcastService $inProgress observable](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/events.md#the-inprogress-observable) in MSAL Angular and [MsalProvider useMsal() hook](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/hooks.md#usemsal-hook) in MSAL React.
+
+In case of concurrent interactive requests, only the first one will initiate an interaction, while the rest will fail with [interaction_in_progress](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/errors.md#interaction_in_progress) error.
+
+### Attempt silent request first, then interactive
+
+When requesting tokens, always use `acquireTokenSilent` first, falling back to interactive token acquisition if needed. Exception to this is when doing token acquisition outside of an app context that does not have access to interaction status or cannot ensure other components making interactive requests. For instance, if you are making use of `acquireTokenSilent` in a utility function,
+
+Multiple concurrent silent requests are permitted. If two or more of silent requests are made at the same time, only one would go to the network (if needed), but all would receive the response, as long as those requests are for the same requests parameters (e.g. scopes).
+
+### Make one token request per resource
+
+You can only request access tokens for one resource at a time (see [resources and scopes](resources-and-scopes.md)). If needed, you can ask user consent to scopes (permissions) required by more than one resource by using the `extraScopesToConsent` parameter in the request object. Access tokens for previously consented scopes can be acquired silently.
```

Comment:
Hi Derisen! thanks for doing this! 
I have some questions remaining around best practices for concurrency. Given that it's common for apps to have multiple data fetches concurrently running which may require different tokens for different resources/apis, how should this be handled?

Our logic to fetch tokens existed in an axios request interceptor. The problem was when tokens became stale, and the silent token fetch failed, given the concurrent nature of data fetch requests, we had several requests being fired off for msal tokens, causing many redirects (and throwing msal errors). I'm wondering how a user of msal could get around this without implementing their own cache for their tokens (the problem being, we don't find out any tokens are stale in the msal cache before firing off these api requests, at which point now many requests have been fired off and we will have too many redirects). I'm wondering if this is a problem other people will run into, and if we could illustrate how people can avoid these issues

Reply:
@1hw7 no worries! From the top of my head, I'm thinking if the first silent request fails and triggers interaction, the newly obtained refresh token would allow other silent requests to succeed... if only the other interceptors would wait for it. Is it not possible to access msal events or interaction status inside the interceptor and have a logic for that? cc @tnorling @sameerag would probably have better ideas

Reply:
Thanks @derisen for the suggestion and your help! the issue we encountered was that the first request would fail, so we would fall into the catch block to move to an interactive fetch for a token (as we thought was correct from the docs). The part i wasn't clear about in the docs was that users would want to ensure this logic is done in such a place and way that even if multiple requests are going on at the same time, you only want to fallback to this interactive method once so it should ideally be in a component rather than somewhere like an axios interceptor. For that reason, @sameerag had suggested we pull the logic out of the axios interceptor.

We ended up implementing our own token cache to ensure we could safely just do this in one spot in the app, but hope there is a better way we could do that